kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,994 @@
+/// this potentially nested schema. returns the number of parquet fields in `fields` (regardless of
+/// if they are selected or not) and reordering information for the requested fields.
+fn get_indices(
+    start_parquet_offset: usize,
+    requested_schema: &Schema,
+    fields: &Fields,
+    mask_indices: &mut Vec<usize>,
+) -> DeltaResult<(usize, Vec<ReorderIndex>)> {
+    let mut found_fields = HashSet::with_capacity(requested_schema.fields.len());
+    let mut reorder_indices = Vec::with_capacity(requested_schema.fields.len());
+    let mut parquet_offset = start_parquet_offset;
+    for (parquet_index, field) in fields.iter().enumerate() {
+        debug!(
+            "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
+            field.name()
+        );
+        match field.data_type() {
+            ArrowDataType::Struct(fields) => {
+                if let Some((index, _, requested_field)) =
+                    requested_schema.fields.get_full(field.name())
+                {
+                    match requested_field.data_type {
+                        DataType::Struct(ref requested_schema) => {
+                            let (parquet_advance, children) = get_indices(
+                                parquet_index + parquet_offset,
+                                requested_schema.as_ref(),
+                                fields,
+                                mask_indices,
+                            )?;
+                            // advance the number of parquet fields, but subtract 1 because the
+                            // struct will be counted by the `enumerate` call but doesn't count as
+                            // an actual index.
+                            parquet_offset += parquet_advance - 1;
+                            // note that we found this field
+                            found_fields.insert(requested_field.name());
+                            // push the child reorder on
+                            reorder_indices.push(ReorderIndex::Child { index, children });
+                        }
+                        _ => {
+                            return Err(Error::unexpected_column_type(field.name()));
+                        }
+                    }
+                } else {
+                    // We're NOT selecting this field, but we still need to update how much we skip
+                    debug!("Skipping over un-selected struct: {}", field.name());
+                    // offset by number of inner fields. subtract one, because the enumerate still
+                    // counts this field
+                    parquet_offset += count_cols(field) - 1;
+                }
+            }
+            ArrowDataType::List(list_field)
+            | ArrowDataType::LargeList(list_field)
+            | ArrowDataType::ListView(list_field) => {
+                if let Some((index, _, requested_field)) =
```

Comment:
Luckily (I think), we don't need to deal with too much about how stuff is actually represented here. We just want to make sure there is _logical_ agreement between the schemas.

We don't have to convert because we get an [`ArrowReaderMetadata`](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html) from the file and then call [`schema`](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderMetadata.html#method.schema) on it, which gives us back an already converted Arrow schema.

Reply:
Discussed offline. Our new casting infra can deal with these cases as they come up. Currently don't anticipate this being needed for lists.